doc/dbus/bus/org.opensuse.Agama.Questions1.LuksActivation.bus.xml

```diff
@@ -1,41 +1,53 @@
+
 <!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
-"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
-<node name="/org/opensuse/Agama/Questions1/2">
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="org.freedesktop.DBus.Introspectable">
+    <method name="Introspect">
+      <arg type="s" direction="out"/>
+    </method>
+  </interface>
+  <interface name="org.freedesktop.DBus.Peer">
+    <method name="Ping">
+    </method>
+    <method name="GetMachineId">
+      <arg type="s" direction="out"/>
+    </method>
+  </interface>
+  <interface name="org.opensuse.Agama.Questions1.Generic">
```

Comment:
*Questions1.Generic* is defined three times (*Questions1.bus.xml*, *Questions1.Generic.bus.xml* and *Questions1.LuksActivation.bus.xml*). Is that expected? I am not sure if that is the best way for documenting objects with dynamic interfaces.

Reply:
In fact, I think the file names could be misleading. For example, I would expect that *doc/dbus/bus/org.opensuse.Agama.Questions1.Generic.bus.xml* defines the *Questions1.Generic* interface, but it defines all the interfaces of an object that only includes *Questions1.Generic* interface.

I would expect to have both, object docs and interface docs. And an object doc only includes the name of the interfaces, for example:

~~~
doc/dbus/objects/question/generic.xml
<node>
  <interface name="org.freedesktop.DBus.Introspectable" />
  <interface name="org.freedesktop.DBus.Peer" />
  <interface name="org.opensuse.Agama.Questions1.Generic" />
</node

doc/dbus/objects/question/luks_activation.xml
<node>
  <interface name="org.freedesktop.DBus.Introspectable" />
  <interface name="org.freedesktop.DBus.Peer" />
  <interface name="org.opensuse.Agama.Questions1.Generic" />
  <interface name="org.opensuse.Agama.Questions1.LuksActivation" />
</node

doc/dbus/interfaces/org.opensuse.Agama.Questions1.Generic.xml
<interface name="org.opensuse.Agama.Questions1.Generic">
  <property name="Answer" type="s" access="readwrite"/>
  <property name="DefaultOption" type="s" access="read"/>
  <property name="Id" type="u" access="read"/>
  <property name="Options" type="as" access="read"/>
  <property name="Text" type="s" access="read"/>
</interface>
~~~

In general, something similar to what [UDisks2](http://storaged.org/doc/udisks2-api/latest/ref-dbus.html) does. What do you think?

Reply:
> I would expect to have both, object docs and interface docs.

Very good point, I have so far concentrated on the automated part and limited the output to the interfaces.
We also need documenting the objects (object trees) and which interfaces they implement. Probably some manually written pieces.